### PR TITLE
Add git-status function and status api in sdk

### PR DESCRIPTION
**Pull request by s8sg**
This commit adds git-status function to handle commit status updates,
along with the below changes
- sdk add status update api to abstract the call to git-status, append
multiple statuses and re-use auth token for consecutive call
- updated sdk is vendor for other functions
- moves status update vendor code from buildshiprun to git-status
- function specific status and overall status for `stack-deploy` by separate contexts 
- pending status for each context

Signed-off-by: Swarvanu Sengupta <swarvanusg@gmail.com>

Fix Issue:
#41  #36

**Overview:** 
- Separate context for each function deploy + context for stack deploy 
- Separate Openfaas Function For Handing Status Update `git-status`
- Using SDK for status update `sdk/statuses.go`
- Using reuse of token (from sdk on a single invocation)
- Using grouping of Status and accept multiple status at `git-status`

**How status gets modified:** 

##### Stack example:
```
provider:
  name: faas
  gateway: http://127.0.0.1:8080

functions:
  regex_go:
    lang: go
    handler: ./regex_go
    image: s8sg/regex_go

  regex_py:
    lang: python
    handler: ./regex_py
    image: s8sg/regex_py
``` 

screen shot 2018-05-02 at 6 03 37 pm

screen shot 2018-05-04 at 12 33 21 pm

screen shot 2018-05-04 at 12 33 28 pm

screen shot 2018-05-04 at 12 33 36 pm

### Links: 
* passed

| context |     url |
| --- | --- |
| stack-deploy | gateway_url: https://7d86c827.ngrok.io/ui/   |
|                        |    pretty url: https://s8sg.7d86c827.ngrok.io     |                    
| regex_go              | gateway url : https://7d86c827.ngrok.io/function/s8sg-regex_go / |
|                        |    pretty url: https://s8sg.7d86c827.ngrok.io/regex_go |
| regex_py              | gateway url : https://7d86c827.ngrok.io/function/s8sg-regex_py / |
|                        |    pretty url: https://s8sg.7d86c827.ngrok.io/regex_py |

* failed/pending

| context |     url |
| --- | --- |
| stack-deploy | git repo   |                    
| regex_go              | git repo |
| regex_py              | git repo |

#### Success
Make some nonbreaking change say `README.md` and it should have the commit progress as described on the PR
> The stack deploy, regex_go and regex_py should be successful   
> overall status must be a success 

#### Failure tar creation
change the handler name in `stack.yml`
```yaml
  regex_py:
    lang: python
    handler: ./regex_invalid
    image: s8sg/regex_py
```
> The stack deploy fails 
> overall status will be a failure

Fix the `stack.yml`

#### Failure one function
change the go code `regex_go/handler.go`, to make it fail deliberately    
remove the import 
```go
	"encoding/json"
```
> In this case the stack deploy will be successful 
> regex_py will be successful
> regex_go will be failed
> overall status will be a failure

#### Nothing should be stuck on Pending
None of the statuses should get stuck on pending, they should either fail or pass

I think this is enough to test the functionalities

You can check the `sdk` validation 

#### Check if the public key is used from provided name
Create a new secret
```bash
docker secret create new-private-key <your_private_key_file>.pem
```
and change the secret name at `github.yml`
```
private_key: new-private-key
```
Redeploy and Commit Non-Breaking Change

#### Check cases when status update fails
Create a new secret
```bash
echo "invalid key" > invalid
docker secret rm new-private-key
docker secret create new-private-key invalid
```
Redeploy and Commit Non-Breaking Change
It will create the error in the buildshiprun or in other function as:
```bash
invalid auth token received, token : ( <error string> )
```

> Github-status return the auth token, as it put it in stdout, the `combine_output` must be disabled to reuse the token

#### Reuse authtoken for multiple calls from same function
from `docker service logs github-status` find `reusing provided auth token`
```bash
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | {"token":"v1.12334db41169034c40a4a8c21c4740a02e4d4558"}
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:38 Forking fprocess.
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:38 Query  
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:38 Path  /function/github-status
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 stderr: 2018/08/06 05:39:39 auth token is created
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Status: success, Context: regex_py, Service: s8sg-regex_go, GitHub AppID: 138863, Repo: regex_go, Owner: s8sg
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Duration: 1.388215 seconds
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | {"token":"v1.2dc649211b6757b2cf7c2591b0b4201e91fd538c"}
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Forking fprocess.
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Query  
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Path  /function/github-status
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | {"token":"v1.12334db41169034c40a4a8c21c4740a02e4d4558"}
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:40 stderr: 2018/08/06 05:39:39 reusing provided auth token
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:39 Status: success, Context: stack-deploy, Service: s8sg-regex_go, GitHub AppID: 138863, Repo: regex_go, Owner: s8sg
github-status.1.r0n1ggmzx15v@linuxkit-025000000001    | 2018/08/06 05:39:40 Duration: 1.000299 seconds
``` 

I have:

- [x] updated the documentation if required
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests
- [x] added test guidelines

### CODE REVIEW

**Comment by s8sg** [comment]
@alexellis @viveksyngh 
It was easier for me to just create a new pr to merge for rebase issue 
It is same as #63 

The net file changes are listed [here](https://gist.github.com/s8sg/98fba880d2a9e805613bc727d505658d)

**Comment by viveksyngh** [comment]
@s8sg This may not be related to your PR, but now when I try to deploy my functions on local it failed to build in builder with error. 

```
Deployment error: Error response from daemon: rpc error: code = InvalidArgument desc = ContainerSpec: "failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c pip install -r requirements.txt]: exit code 127" is not a valid repository/tag
```

Function status gets updated in github but is wrong because function deployment was failed but it was updated as successful.

Note: Go function gets deployed successfully and status update is also correct.

**Comment by viveksyngh** [comment]
'Success` and `Failure tar creation` Test cases seem to be working fine.

`Failure one function` test case is also failing. See the status here. Commit with the message `make it fail`. https://github.com/viveksyngh/regex_go/commits/master

**Comment by viveksyngh** [comment]
**Check if the public key is used from provided name**
Seems to be failing for me. I created the private key with a new name but I get the error `invalid auth token received, token : ( exit status 1 ),`. I think this can be the issue with the mount path for secrets ?

**Check cases when status update fails** 

Seems to be working


**Check cases when status update fails**

Seems to be working

**Comment by s8sg** [comment]
@viveksyngh Thanks for the detailed report. 

**Failure one function** 
I checked and I agree the test case is failing 
The problem here is not the changes related to the `github-status` changes but the failure check in buildshiprun code.   
In the log of the `buildshiprun`
```bash
buildshiprun.1.l81fyjakncif@linuxkit-025000000001    | 2018/07/02 02:05:13 buildshiprun: image 'failed to solve: rpc error: code = Unknown desc = exit code 2'
buildshiprun.1.l81fyjakncif@linuxkit-025000000001    | 2018/07/02 02:05:13 Deploying failed to solve: rpc error: code = Unknown desc = exit code 2 as s8sg-regex_go
buildshiprun.1.l81fyjakncif@linuxkit-025000000001    | functionExists status: 200 OK
buildshiprun.1.l81fyjakncif@linuxkit-025000000001    | Deploying: failed to solve: rpc error: code = Unknown desc = exit code 2 as s8sg-regex_go
buildshiprun.1.l81fyjakncif@linuxkit-025000000001    | Deploy status: 400 Bad Request
```
The **initial problem** here is:    
The image name here is `invalid`. In the code (`buildshiprun/handler.go`) we are just checking:
```go
        if strings.Contains(imageName, "exit status") == true {
                msg := "Unable to build image, check builder logs"
                status.AddStatus(sdk.Failure, msg, sdk.FunctionContext(event.Service))
                reportStatus(status)
                log.Fatal(msg)
                auditEvent.Message = fmt.Sprintf("buildshiprun failure: %s", msg)
                sdk.PostAudit(auditEvent)
                return msg
        }
```
Which is not enough. 

The **2nd problem** we need to check the resp Status in `deployFunction`. In the log it clearly got a `404` but as the code doesn't check the resp status, it reports no error.
```go
     res, err = c.Do(httpReq)

        if err != nil {
                fmt.Println(err)
                return "", err
        }

        defer res.Body.Close()
        fmt.Println("Deploy status: " + res.Status)
```
> This is not an issue that falls under the current PR but we need a separate PR to update the Current Failure check code

**Check if the public key is used from provided name**
It seems the the `gateway_config.yml` and `github.yml` both have the:
```yaml
private_key: 
```
And its being used only in `github-status`.
Although as it is a `github` specific option it might need to be in `github.yml` only

> I'll commit by removing it from the `gateway_config.yml`

**Comment by s8sg** [comment]
@viveksyngh I have created a new issue related to **proper error check** at [issue-84](https://github.com/openfaas/openfaas-cloud/issues/84)  and Implemented the fix and created a PR #85

**Comment by alexellis** [comment]
Perhaps the token re-use should also be structured JSON

File: buildshiprun/handler.go [hunk 1]
```diff
@@ -108,24 +105,27 @@ func Handle(req []byte) string {
 			Network: "func_functions",
 			Labels: map[string]string{
```
Comment (alexellis) [inline buildshiprun/handler.go]: We will need to start storing some of the data in annotations - this is a new feature coming in via faas-netes / faas-swarm etc. IT allows for any value including a date/time stamp in human-readable format.

File: buildshiprun/handler.go [hunk 2]
```diff
@@ -269,89 +217,68 @@ func enableStatusReporting() bool {
...
-		return
-	}
+	// return repositoryURL + imageName[strings.Index(imageName, "/"):]
+}
 
-	if token == "" {
-		fmt.Printf("failed to report status %v, error: authentication failed Invalid token\n", repoStatus)
-		return
-	}
+// function to build Event from Environment Variable
```
Comment (alexellis) [inline buildshiprun/handler.go]: Comment should be prefixed with `// BuildEventFromEnv `

File: buildshiprun/handler.go [hunk 3]
```diff
@@ -269,89 +217,68 @@ func enableStatusReporting() bool {
...
 
-func getPrivateKey() string {
-	// we are taking the secrets name from the env, by default it is fixed
-	// to private_key.pem.
-	// Although user can make the secret with a specific name and provide
-	// it in the stack.yaml and also specify the secret name in github.yml
-	privateKeyName := os.Getenv("private_key")
-	if privateKeyName == "" {
-		privateKeyName = defaultPrivateKeyName
+func GetEnv(service string) map[string]string {
```
Comment (alexellis) [inline buildshiprun/handler.go]: Should these now be private? `getEnv` etc

File: buildshiprun/handler.go [hunk 4]
```diff
@@ -269,89 +217,68 @@ func enableStatusReporting() bool {
...
-	return &github.RepoStatus{State: &status, TargetURL: &url, Description: &desc, Context: &context}
+	return envVars
 }
 
-func getImageName(repositoryURL, pushRepositoryURL, imageName string) string {
-
-	return strings.Replace(imageName, pushRepositoryURL, repositoryURL, 1)
-
-	// return repositoryURL + imageName[strings.Index(imageName, "/"):]
+func GetSecret(owner, service string) []string {
```
Comment (alexellis) [inline buildshiprun/handler.go]: Upper-case letter denotes a public method, should it be lower-case?

Is there a corresponding unit test for this?

File: sdk/events.go [hunk 5]
```diff
@@ -1,5 +1,7 @@
 package sdk
 
+import ()
```
Comment (alexellis) [inline sdk/events.go]: Is this line needed?

File: sdk/events.go [hunk 6]
```diff
@@ -14,6 +16,33 @@ type PushEvent struct {
...
+	Image          string            `json:"image"`
+	Sha            string            `json:"sha"`
+	URL            string            `json:"url"`
+	InstallationID int               `json:"installationID"`
+	Environment    map[string]string `json:"environment"`
+	Secrets        []string          `json:"secrets"`
+}
+
+// function to build Event from PushEvent
+func BuildEventFromPushEvent(pushEvent PushEvent) *Event {
```
Comment (alexellis) [inline sdk/events.go]: Write comment like:

```
// BuildEventFromPushEvent .... 
```

File: sdk/events.go [hunk 7]
```diff
@@ -14,6 +16,33 @@ type PushEvent struct {
 	}
 	AfterCommitID string `json:"after"`
 	Installation  struct {
-		Id int `json:"id"`
+		ID int `json:"id"`
 	}
 }
+
+// EventInfo to pass/store events across functions
+type Event struct {
```
Comment (alexellis) [inline sdk/events.go]: Write comment with a prefix of the object name `// Event .... `

File: sdk/statuses.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
+package sdk
```
Comment (alexellis) [inline sdk/statuses.go]: Perhaps this should be singular sdk/status.go ?

**Comment by alexellis** [comment]
All functions that talk to the Gateway, will need to add the secret in their list of secrets in stack.yml and also we will need to test this work before merging with basic auth enabled on the gateway.

**Comment by alexellis** [comment]
Does this address the use-case where the user doesn't give stack.yml but my-function.yml instead? Do they get an error on the commit?

**Comment by alexellis** [comment]
Please try to cover the new code with unit tests where applicable.

**Comment by alexellis** [comment]
Comments should follow the Golang syntax which is:

```
// nameOfMethod etc
func nameOfMethod() {

}

// myVar ....
var myVar string
```

File: buildshiprun/handler.go [hunk 9]
```diff
@@ -269,89 +217,68 @@ func enableStatusReporting() bool {
...
+// function to build Event from Environment Variable
+func BuildEventFromEnv() (*sdk.Event, error) {
+	var err error
+	info := sdk.Event{}
 
-	client := auth.MakeClient(ctx, token)
+	info.Service = os.Getenv("Http_Service")
+	info.Owner = os.Getenv("Http_Owner")
+	info.Repository = os.Getenv("Http_Repo")
+	info.Sha = os.Getenv("Http_Sha")
```
Comment (alexellis) [inline buildshiprun/handler.go]: Could that be info.SHA?

File: sdk/statuses.go [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+// Status to post github-status to git-status function
+type Status struct {
+	CommitStatuses map[string]CommitStatus `json:"commit-statuses"`
+	EventInfo      Event                   `json:"event"`
+	AuthToken      string                  `json:"auth-token"`
+}
+
+// builds a status
```
Comment (alexellis) [inline sdk/statuses.go]: `// BuildStatus creates a build status.`

File: git-tar/function/handler.go [hunk 11]
```diff
@@ -26,29 +26,40 @@ func Handle(req []byte) []byte {
 		os.Exit(-1)
 	}
 
+	statusEvent := sdk.BuildEventFromPushEvent(pushEvent)
+	status := sdk.BuildStatus(statusEvent, "")
```
Comment (alexellis) [inline git-tar/function/handler.go]: Create a const called `EmptyAuthToken` and pass this instead, for readability.

**Comment by s8sg** [comment]
@alexellis I need some help to rebase the branch and merge

**Comment by s8sg** [comment]
@alexellis @viveksyngh This branch has rebased and all the review comments have been updated.

File: buildshiprun/handler.go [hunk 12]
```diff
@@ -55,9 +52,10 @@ func Handle(req []byte) string {
 
 	if err != nil {
 		fmt.Println(err)
-		reportStatus("failure", err.Error(), "BUILD", event)
 		auditEvent.Message = fmt.Sprintf("buildshiprun failure: %s", err.Error())
 		sdk.PostAudit(auditEvent)
+		status.AddStatus(sdk.Failure, err.Error(), sdk.FunctionContext(event.Service))
```
[the remark this answers is not in the document]
Comment (s8sg) [reply buildshiprun/handler.go]: I'll update that

File: sdk/status.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+	Failure = "failure"
+	Pending = "pending"
+)
+
+// context constant
+const (
+	Deploy         = "%s"
+	Stack          = "stack-deploy"
+	EmptyAuthToken = ""
+	tokenString    = "token"
```
[the remark this answers is not in the document]
Comment (s8sg) [reply sdk/status.go]: I'll update

File: sdk/status.go [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+	match := validToken.FindString(token)
+	// token should be the whole string
+	if len(match) == len(token) {
+		return true
+	}
+	return false
+}
+
+// MarshalToken mardhal a token into json
+func MarshalToken(token string) string {
```
Comment (alexellis) [inline sdk/status.go]: I'm confused what this method does? it starts as a string and ends as a string?

File: github-status/handler_test.go [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+}
+
+func TestTokenMarshal(t *testing.T) {
+	testTokens := map[string]string{
+		"token with space":                                sdk.EmptyAuthToken,
+		"token$With@special=Char":                         sdk.EmptyAuthToken,
+		"v1.afbce39asdasd8be30123317cef123321ae991cf40f7": "v1.afbce39asdasd8be30123317cef123321ae991cf40f7",
+		"token=v1.afbce39asdasd8be30123":                  sdk.EmptyAuthToken,
+		" ": sdk.EmptyAuthToken}
+	for token, result := range testTokens {
```
Comment (alexellis) [inline github-status/handler_test.go]: Use sub-tests here.

File: github-status/handler.go [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// marshal token
+	token = sdk.MarshalToken(token)
+
+	// return auth token so that it can be reused form a same function
+	return token
+}
+
+func buildPublicStatusURL(status string, statusContext string, event *sdk.Event) string {
```
[the remark this answers is not in the document]
Comment (s8sg) [reply github-status/handler.go]: Theres already test-cases are present, I have not added any additional check as the code is same

File: buildshiprun/handler.go [hunk 17]
```diff
@@ -161,19 +161,19 @@ func getReadOnlyRootFS() bool {
 	return readOnly
 }
 
-func getEvent() (*eventInfo, error) {
+func getEvent() (*sdk.Event, error) {
```
[the remark this answers is not in the document]
Comment (s8sg) [reply buildshiprun/handler.go]: There's no extra filed added in the Event expect that it's moved to `sdk`. No additional test cases added

File: sdk/status.go [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+	marshalToken, _ := json.Marshal(map[string]string{tokenString: token})
+	return string(marshalToken)
+}
+
+// UnmarshalToken unmarshal a token and validate
+func UnmarshalToken(data []byte) string {
```
Comment (alexellis) [inline sdk/status.go]: This signature should return `(string, error)` and not swallow or log the message to the console. That's for the calling code to decide.

File: sdk/status.go [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+func ValidToken(token string) bool {
+	match := validToken.FindString(token)
+	// token should be the whole string
+	if len(match) == len(token) {
+		return true
+	}
+	return false
+}
+
+// MarshalToken mardhal a token into json
```
[the remark this answers is not in the document]
Comment (s8sg) [reply sdk/status.go]: I'll update it

File: sdk/status.go (hunk 14 above)
Comment (s8sg) [reply sdk/status.go]: It marshal the token in a `json`. Its used by the `github-status`

File: sdk/status.go [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+	status := Status{}
+	status.EventInfo = *event
+	status.CommitStatuses = make(map[string]CommitStatus)
+	status.AuthToken = token
+
+	return &status
+}
+
+// UnmarshalStatus unmarshal a status object from json
+func UnmarshalStatus(data []byte) (*Status, error) {
```
Comment (alexellis) [inline sdk/status.go]: Please add a commit for a unit test for `UnmarshalStatus` including a test where an error is generated due to a bad input.

Comment (s8sg) [reply sdk/status.go]: I have added a separate Test case to check the `UnmarshalStatus` Error cases, any other cases are getting covered here.

File: github-status/handler_test.go (hunk 15 above)
Comment (s8sg) [reply github-status/handler_test.go]: I have added a separate Test case to check the `UnmarshalStatus` Error cases, any other scenario are getting covered here.

Comment (alexellis) [reply github-status/handler_test.go]: You need to use a sub-test, it's something like calling `t.Test()` and passing in a scenario name. Please refer to test tables too. @rgee0 can point you in the right direction.

Comment (rgee0) [reply github-status/handler_test.go]: Pop over to [Derek](https://github.com/alexellis/derek/blob/master/commentHandler_test.go) and have a look at the tests in [commentHandler](https://github.com/alexellis/derek/blob/master/commentHandler_test.go).

Basically, put all your test cases into a struct then loop through that to test each scenario.  Using a test.title helps in tracking down any fails.

Comment (s8sg) [reply github-status/handler_test.go]: Updated in the new commit
https://github.com/openfaas/openfaas-cloud/pull/83/files#diff-b3c2497a977c4257af245d1f1f78474bR82

File: github-status/handler_test.go [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+	val := buildPublicStatusURL("not-supported", sdk.BuildFunctionContext(event.Service), event)
+	want := event.URL
+
+	if val != want {
+		t.Errorf("building PublicURL: want %s, got %s", want, val)
+		t.Fail()
+	}
+}
+
+func TestToken(t *testing.T) {
```
Comment (s8sg) [inline github-status/handler_test.go]: The new commit adds the sub tests with structured data set

File: sdk/status.go (hunk 20 above)
Comment (s8sg) [reply sdk/status.go]: I added a Separated test case for Error Check

File: buildshiprun/handler_test.go [hunk 22]
```diff
@@ -165,12 +95,51 @@ func Test_GetImageName(t *testing.T) {
 
 func Test_ValidImage(t *testing.T) {
```
Comment (s8sg) [inline buildshiprun/handler_test.go]: I have tested the regex for the below cases [here](https://regex101.com/r/wM3Svf/1)

**Comment by s8sg** [comment]
screenshot 2018-08-07 at 2 04 11 pm

I performed the above tests to cover multiple scenario
1. No `stack.yml`
2. Invalid Handler
3. Error in One function
4. Success

**Pull request merged**